uefi/firmwarevolume.go

```diff
@@ -189,6 +190,39 @@ func (fv *FirmwareVolume) Assemble() ([]byte, error) {
 	if err != nil {
 		return nil, err
 	}
+	// Make sure we don't go over the size.
+	fvLen := uint64(len(fv.buf))
+	fOffset := fv.DataOffset
+	for _, f := range fv.Files {
+		fBuf, err := f.Assemble()
+		if err != nil {
+			return nil, err
+		}
+		fLen := uint64(len(fBuf))
+		if fLen+fOffset > fvLen {
+			// TODO: Actually loop through and calculate the full size so we know how much to reduce by.
+			// For now we just return early
+			return nil, errors.New("insufficient space in FV, files too big")
+		}
+		// Overwrite old data in the firmware volume.
+		// We have to pad to the 8 byte alignments.
+		for alignBytes := align8(fOffset) - fOffset; alignBytes > 0; alignBytes-- {
+			fv.buf[fOffset] = 0xFF
+			fOffset++
+		}
+		copy(fv.buf[fOffset:], fBuf)
+		fOffset += fLen
+	}
+
+	if fOffset == fvLen {
+		// We're done here, no need to fill the rest of the fv
+		return fv.buf, nil
+	}
+
+	// Fill to the end with FFs
+	for ; fOffset < fvLen; fOffset++ {
+		fv.buf[fOffset] = 0xFF
```

Comment:
Since you're filling with 0xFF in multiple places, you can add a function:

```
func fillFF(data []byte) {
	for i := range data {
		data[i] = 0xff
	}
}
```

And call like this:

```
fillFF(fv.buf[fOffset:fvLen])
```

And on line 209, you can use:

```
fillFF(fv.buf[fOffset:align8(fOffset)])
fOffset = align8(fOffset)
```

Reply:
I could, but I just realized that for the alignments, I have to check if the filling of FFs will go out of bounds. I'll redo it slightly differently

Reply:
done